rig-core/src/providers/gemini/completion.rs

```diff
@@ -0,0 +1,607 @@
+pub struct GenerateContentRequest {
+    pub contents: Vec<Content>,
+    pub tools: Option<Vec<Tool>>,
+    pub tool_config: Option<ToolConfig>,
+    pub generation_config: Option<GenerationConfig>,
+    pub safety_settings: Option<Vec<SafetySetting>>,
+    pub system_instruction: Option<String>,
+    // cachedContent: Optional<String>
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Tool {
+    pub function_declaration: FunctionDeclaration,
+    pub code_execution: Option<CodeExecution>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FunctionDeclaration {
+    pub name: String,
+    pub description: String,
+    pub parameters: Option<Vec<Schema>>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ToolConfig {
+    pub schema: Option<Schema>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CodeExecution {}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SafetySetting {
+    pub category: HarmCategory,
+    pub threshold: HarmBlockThreshold,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum HarmBlockThreshold {
+    HarmBlockThresholdUnspecified,
+    BlockLowAndAbove,
+    BlockMediumAndAbove,
+    BlockOnlyHigh,
+    BlockNone,
+    Off,
+}
+
+// =================================================================
+// Rig Implementation Types
+// =================================================================
+
+#[derive(Clone)]
+pub struct CompletionModel {
+    client: Client,
+    pub model: String,
+}
+
+impl CompletionModel {
+    pub fn new(client: Client, model: &str) -> Self {
+        Self {
+            client,
+            model: model.to_string(),
+        }
+    }
+}
+
+impl completion::CompletionModel for CompletionModel {
+    type Response = GenerateContentResponse;
+
+    async fn completion(
+        &self,
+        mut completion_request: CompletionRequest,
+    ) -> Result<completion::CompletionResponse<GenerateContentResponse>, CompletionError> {
+        // QUESTION: Why do Anthropic/openAi implementation differ here? OpenAI adds the preamble but Anthropic does not.
+
+        let mut full_history = if let Some(preamble) = &completion_request.preamble {
+            vec![completion::Message {
+                role: "system".into(),
+                content: preamble.clone(),
+            }]
+        } else {
+            vec![]
+        };
+
+        full_history.append(&mut completion_request.chat_history);
+
+        let prompt_with_context = completion_request.prompt_with_context();
+
+        full_history.push(completion::Message {
+            role: "user".into(),
+            content: prompt_with_context,
+        });
+
+        // Handle Gemini specific parameters
+        let mut generation_config =
+            GenerationConfig::try_from(completion_request.additional_params.unwrap_or_default())?;
```

Comment:
This is interesting. One negative of this is that if Gemini adds new fields and rig isn't updated, you won't be able to use them since this struct will ignore them. I chose not to do this for Anthropic.

Reply:
@cvauclair What do you think? I thought that from a dev UX perspective having access to types/docs definitions was useful. IMO field updates/additions should be *relatively* infrequent so that we can wait for someone to raise a repo issue to implement the change.

Reply:
This isn't a case of rare and breaking changes, new fields being added would _not_ warrant a new version of the API (it would just be added) yet they would cause this model to break. It's not that infrequent as the providers match each other's capabilities (see OpenAI prompt caching) and it also forces users to wait for a rig update unable to use the newest features of their provider.

This might be a fine compromise, but I want to make it clear that this is the consequence.